programs/solana/programs/ics07-tendermint/src/instructions/update_client.rs

```diff
@@ -197,71 +202,910 @@ fn handle_consensus_state_storage<'info>(
+    }
+
+    #[test]
+    fn test_update_client_with_wrong_trusted_height() {
+        let fixture = load_happy_path_fixture();
+        let update_message_fixture = fixture.update_client_message;
+        let client_message = hex_to_bytes(&update_message_fixture.client_message_hex);
+        let new_height = update_message_fixture.new_height;
+
+        let scenario = setup_update_client_test_scenario(client_message.clone(), new_height, None);
+        let mut accounts = scenario.accounts;
+
+        // Use wrong trusted consensus state (height differs from fixture)
+        let wrong_trusted_height = if update_message_fixture.trusted_height > 1 {
+            update_message_fixture.trusted_height - 1
+        } else {
+            update_message_fixture.trusted_height + 10
+        };
+
+        let (wrong_trusted_consensus_state_pda, _) = Pubkey::find_program_address(
+            &[
+                b"consensus_state",
+                scenario.client_state_pda.as_ref(),
+                &wrong_trusted_height.to_le_bytes(),
+            ],
+            &crate::ID,
+        );
+
+        let (new_consensus_state_pda, _) = Pubkey::find_program_address(
+            &[
+                b"consensus_state",
+                scenario.client_state_pda.as_ref(),
+                &new_height.to_le_bytes(),
+            ],
+            &crate::ID,
+        );
+
+        // Create instruction with wrong trusted consensus state
+        let instruction = create_update_client_instruction(
+            scenario.client_state_pda,
+            wrong_trusted_consensus_state_pda, // Wrong trusted state
+            new_consensus_state_pda,
+            scenario.payer,
+            client_message,
+        );
+
+        // Add empty account for wrong trusted state
+        accounts.push((
+            wrong_trusted_consensus_state_pda,
+            create_empty_consensus_state_account(),
+        ));
+
+        let result = execute_update_client_instruction(&instruction, &accounts);
+        // Should fail because account validation fails for wrong PDA
+        assert_error_code(
+            result,
+            ErrorCode::AccountValidationFailed,
+            "Wrong trusted height",
+        );
+    }
+
+    #[test]
+    fn test_update_client_with_expired_header() {
+        let scenario = setup_happy_path_test_scenario();
+
+        // Use a clock time that's way in the future to make header appear expired
+        let future_timestamp =
+            get_expired_clock_timestamp_for_header(&scenario.update_message_fixture);
+        let (clock_pubkey, clock_account) = create_clock_account(future_timestamp);
+
+        // Replace the clock account
+        let mut accounts = scenario.accounts;
+        for (pubkey, account) in accounts.iter_mut() {
+            if pubkey == &clock_pubkey {
+                *account = clock_account;
+                break;
+            }
+        }
```

Comment:
Maybe more idiomatic approach could be used like:

```suggestion
        if let Some((_, acc)) = accounts.iter_mut().find(|(key, _)| *key == clock_pubkey) {
           *acc = clock_account;
        }
```